### PR TITLE
Rector: CQ - SimplifyIfReturnBoolRector

### PR DESCRIPTION
**Pull request by sreichel**
See:
- https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#simplifyifreturnboolrector

Rector; 
- added config file
- added to workflow

### CODE REVIEW

File: app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php [hunk 1]
```diff
@@ -53,13 +53,8 @@ public function isShow()
...
         }
-        if ($this->getCriticalCount() == 0 && $this->getMajorCount() == 0 && $this->getMinorCount() == 0
-            && $this->getNoticeCount() == 0
-        ) {
-            return false;
-        }
-
-        return true;
+        return !($this->getCriticalCount() == 0 && $this->getMajorCount() == 0 && $this->getMinorCount() == 0
+            && $this->getNoticeCount() == 0);
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php:57]: We can simplify with DeMorgan's Law `!(A && B) == (!A || !B)`:
```suggestion
        return $this->getCriticalCount() || $this->getMajorCount() || $this->getMinorCount() || $this->getNoticeCount();
```

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php:57]: Step by step ... :)

https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#simplifydemorganbinaryrector

File: app/code/core/Mage/Adminhtml/Block/Sales/Items/Abstract.php [hunk 2]
```diff
@@ -562,10 +562,7 @@ public function canShipPartially($order = null)
             $order = Mage::registry('current_shipment')->getOrder();
         }
         $value = $order->getCanShipPartially();
-        if (!is_null($value) && !$value) {
-            return false;
-        }
-        return true;
+        return !(!is_null($value) && !$value);
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Sales/Items/Abstract.php:565]: DeMorgan's Law `!(A && B) == (!A || !B)`:

```suggestion
        return is_null($value) || $value;
```

Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Sales/Items/Abstract.php:565]: DeMorgan's Law `!(A && B) == (!A || !B)`:

```suggestion
        return is_null($value) || $value;
```

File: app/code/core/Mage/Adminhtml/Block/Sales/Items/Abstract.php [hunk 3]
```diff
@@ -580,17 +577,11 @@ public function canShipPartiallyItem($order = null)
             $order = Mage::registry('current_shipment')->getOrder();
         }
         $value = $order->getCanShipPartiallyItem();
-        if (!is_null($value) && !$value) {
-            return false;
-        }
-        return true;
+        return !(!is_null($value) && !$value);
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Sales/Items/Abstract.php:580]: DeMorgan's Law `!(A && B) == (!A || !B)`:

```suggestion
        return is_null($value) || $value;
```

File: app/code/core/Mage/Adminhtml/Block/Sales/Items/Abstract.php [hunk 4]
```diff
@@ -580,17 +577,11 @@ public function canShipPartiallyItem($order = null)
...
     }
 
     public function isShipmentRegular()
     {
-        if (!$this->canShipPartiallyItem() || !$this->canShipPartially()) {
-            return false;
-        }
-        return true;
+        return !(!$this->canShipPartiallyItem() || !$this->canShipPartially());
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Sales/Items/Abstract.php:585]: Apply DeMorgan's Law: `!(A || B) == (!A && !B)`
```suggestion
        return $this->canShipPartiallyItem()  && $this->canShipPartially();
```

File: app/code/core/Mage/Adminhtml/Block/Sales/Order/Invoice/Create/Form.php [hunk 5]
```diff
@@ -102,10 +102,7 @@ public function hasInvoiceShipmentTypeMismatch()
     public function canShipPartiallyItem()
     {
         $value = $this->getOrder()->getCanShipPartiallyItem();
-        if (!is_null($value) && !$value) {
-            return false;
-        }
-        return true;
+        return !(!is_null($value) && !$value);
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Sales/Order/Invoice/Create/Form.php:105]: DeMorgan's Law `!(A && B) == (!A || !B)`:

```suggestion
        return is_null($value) || $value;
```

File: app/code/core/Mage/Adminhtml/Controller/Action.php [hunk 6]
```diff
@@ -395,13 +395,8 @@ protected function _validateSecretKey()
...
         }
-
-        if (!($secretKey = $this->getRequest()->getParam(Mage_Adminhtml_Model_Url::SECRET_KEY_PARAM_NAME, null))
-            || !hash_equals(Mage::getSingleton('adminhtml/url')->getSecretKey(), $secretKey)
-        ) {
-            return false;
-        }
-        return true;
+        return !(!($secretKey = $this->getRequest()->getParam(Mage_Adminhtml_Model_Url::SECRET_KEY_PARAM_NAME, null))
+            || !hash_equals(Mage::getSingleton('adminhtml/url')->getSecretKey(), $secretKey));
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Controller/Action.php]:
```suggestion
        return ($secretKey = $this->getRequest()->getParam(Mage_Adminhtml_Model_Url::SECRET_KEY_PARAM_NAME, null))
            && hash_equals(Mage::getSingleton('adminhtml/url')->getSecretKey(), $secretKey);
```

File: app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php (hunk 1 above)
Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php:57]: Oh...so rector rules cannot be chained?

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php:57]: With rulessets (chained rules) the PRs would grow to large. Mhhh.

File: app/code/core/Mage/Adminhtml/Controller/Action.php (hunk 6 above)
Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Controller/Action.php]: Continue on the conversation on chaining rector rules, I don't like the idea that it requires another PR to clean up the code above. This example from rector is really bad. It should be clean up before committing. Don't you think?

File: app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php (hunk 1 above)
Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php:57]: Mhhh .... https://getrector.com/demo/ec03b365-db13-4e75-8ecd-d3773ae96870

Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Block/Notification/Toolbar.php:57]: I created an issue: https://github.com/rectorphp/rector/issues/8781

File: app/code/core/Mage/Adminhtml/Controller/Action.php (hunk 6 above)
Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Controller/Action.php]: I'd prefer to revert last commit and add the DeMorgans rule separately.

Let's keep it simple. It only affects a few files.

Maybe add a new issue to not forget about it?

Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Controller/Action.php:399]: I tried to apply the DeMorgan's rule to `!(!is_null($value) && !$value)`, not sure if I made the test correctly here: https://getrector.com/demo/6a07c84e-3a20-45c6-9513-d20329994eb8

Did I make a mistake in the test? It doesn't work, no code change.  If the above test is correct and if this PR is committed, there is no guarantee the convoluted code can be cleaned up by the rule later. So, now what?

Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Controller/Action.php:399]: OK, so rector won't fix the convoluted bool expression now:
image

So, should we apply DeMorgan's manually?

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Controller/Action.php:399]: https://github.com/OpenMage/magento-lts/pull/4152/commits/392d0a88c2a8c68854106c79288e727fa2c1660c

No. I'd take it as it is and apply that rule later. I dont think its worth to change it manually, same pattern is already there.

**Comment by sreichel** [comment]
Issues fixed in #4204

**Comment by sreichel** [comment]
@kiatng can we continue w/o DeMorgan's Law?

**Comment by kiatng** [comment]
I cannot give my approval because in my opinion, some changes are worse than present.

**Comment by sreichel** [comment]
B/c of 6 out of +100 improvements?

Dont get me wrong, i agree with your suggestions, but i dont want to mix it up with automatic fixes.

How about to merge and verify manual changes again?

**Comment by sreichel** [comment]
... at the other hand ... we can leave it open. With rector added as dependency we can skip it for now.

**Comment by kiatng** [comment]
Or submit a PR to improve Rector (which I cannot do).

**Comment by sreichel** [comment]
I'd really like to do, but i am not so familiar with rectors internal functionality.

I think i can write simple custom rules, but in this case ... there are a lot of possibly combinations of "&&", "||" that makes it difficult to make it safe for a general rule. I guess thats the reason why your issue has been closed.

**Comment by fballiano** [comment]
this PR makes the code much harder to read, the benefit is only for the machine, not for the humans.

**Comment by sreichel** [comment]
Disagree. Most of the changes look better to me.

- it safes a lot of lines
- how often do you look at core code?

**Comment by sreichel** [comment]
It's a negated true/fails expression?

So what?

**Comment by fballiano** [comment]
because I'm stupid I think that a negated OR is far less readable. but better make the code 1 line less but much harder to read so that only geniuses will take the time to read it.

again, the code is not for the machines, it's for the people, it should be understandable easily for the people.

only because rector has that rule it doesn't mean it should be applied.

**Comment by sreichel** [comment]
I did not say you're stupid ... you're not.

I'd like to use rector to unify code.

For core, for extensions ...

**Comment by kiatng** [comment]
I need to call out @sreichel comment:

> how often do you look at core code?

Even though it is a simple question, it is irrelevant. The question carries with it negative implications, which are not within standards of our [CoC](https://github.com/OpenMage/magento-lts?tab=coc-ov-file):

> ... we as contributors and maintainers pledge to making participation in our project and our community a harassment-free experience for everyone, regardless of ... level of experience, education, ...

and in particular the second standard:

> Being respectful of differing viewpoints and experiences

Note that our CoC was based on v1.4. The latest version is [2.1](https://www.contributor-covenant.org/version/2/1/code_of_conduct/). I encourage everyone here take the time to read it and think about it. The second standard in this version includes "opinions":

> Being respectful of differing opinions, viewpoints, and experiences

I urge all of us use language, English as well as PHP, that complies to the best standards.

**Comment by sreichel** [comment]
There was no "negative implications"!

However ...

**Comment by sreichel** [comment]
> I need to call out @sreichel comment:
> 
> > how often do you look at core code?

Maybe you got me wrong?

This PR would have added 7 negated return statements.

My question was how often you face these seven particular lines during development,

When you serach the codebase for `return !` you'll have +700 matches, so it is a very common pattern.  Decline to add 7 more b/c of "readablility" sound bit weired.

**Pull request closed without merging**